### ISSUE TITLE
[Debian] Viking 1.10 segfault when quiting

### ISSUE THREAD

**Issue by paulgevers** [viking-gps/viking#202]
I just received Debian bug report [#1055624: viking: Segmentation fault when quitting viking](https://bugs.debian.org/1055624). Quoted here for completeness.  Please let me know if I should ask the reporter for more feedback, or if you believe this is something that should be fixed on the Debian side.

> Package: viking
> Version: 1.10-2
> Severity: important
> 
> I got a segmentation fault when I quit viking:
> 
> [...]
> (viking:45727): GLib-GObject-WARNING **: 00:24:41.116: invalid unclassed pointer in cast to 'VikLayer'
> 
> (viking:45727): GLib-GObject-WARNING **: 00:24:41.116: invalid unclassed pointer in cast to 'GtkWidget'
> 
> (viking:45727): Gtk-CRITICAL **: 00:24:41.116: gtk_widget_get_toplevel: assertion 'GTK_IS_WIDGET (widget)' failed
> zsh: segmentation fault (core dumped)  viking velov-7008-7031-8.gpx
> 
> In gdb:
> 
> [...]
> Core was generated by `viking velov-7008-7031-8.gpx'.
> Program terminated with signal SIGSEGV, Segmentation fault.
> #0  unlink_chunk (p=p@entry=0x55e7b438b860, av=0x7f011f06fc60 <main_arena>)
>     at ./malloc/malloc.c:1628
> 1628    ./malloc/malloc.c: No such file or directory.
> [Current thread is 1 (Thread 0x7f011c3d5180 (LWP 45727))]
> 
> I'ved attached the full backtrace.

[gdb.txt](https://github.com/viking-gps/viking/files/13305228/gdb.txt)

**Comment by rnorris** [comment]
NB I edited your comment to fix the Debian bug link incorporating a spurious ']' - :)

From the stack trace, I note there appeared to be a thread (thread 3) using curl - this implying there is some background task probably downloading map tiles. 

In the past there have been occasional problems if these background tasks were (specifically manually requested to be) cancelled (hopefully all solved now...)
Potentially there still could be some issue at program closure too.
Unfortunately debugging this can be tricky as if the remote services are fast (typically they are) , thus it can be hard to exit the program whilst it is downloading and/or at the exact time doing something that might trigger this issue**.
However as the program is exiting, it's not so critical that it crashes as the program is ending any way - albeit annoying especially if such seg faults get shown to the user.

However I don't quite understand thread 1 - where sqlite3 comes into the equation - but probably this just a side effect of some memory elsewhere going astray.

But I don't think it is a Debian specific problem. 

**NB I use Debian Sid myself and the system installed Viking seems OK and hasn't/doesn't (in any repeatable manner) segfault when exiting during downloading.

I will continue to look into it.

**Comment by vinc17fr** [comment]
I got the segmentation fault and submitted the Debian bug report. And yes, I think that viking was still downloading map tiles. I sometimes use viking on a full 4K screen, so it is easy to get more than 100 tiles in the queue. This is a new machine, so that many tiles are not in the cache yet.

**Comment by vinc17fr** [comment]
I've just reproduced the segmentation fault when quitting viking. This time:
```
** viking [Warning]): curl_download_uri: curl error: 42 for uri http://tile.openstreetmap.org/15/16820/11687.png
** viking [Warning]): Download error: /home/vinc17/.cache/viking/t13s2z0/16820/11687

(viking:102231): GLib-GObject-WARNING **: 23:53:41.521: invalid unclassed pointer in cast to 'VikLayer'

(viking:102231): GLib-GObject-WARNING **: 23:53:41.521: invalid unclassed pointer in cast to 'GtkWidget'

(viking:102231): Gtk-CRITICAL **: 23:53:41.521: gtk_widget_get_toplevel: assertion 'GTK_IS_WIDGET (widget)' failed
```

**Comment by rnorris** [comment]
@vinc17fr - Thanks. However I also have occasionally got the same messages, but no segfault. So not necessarily the definitive cause.
NB curl 'error' of 42 is expected as that is for the download being stopped by the exit request - see https://curl.se/libcurl/c/libcurl-errors.html - of course Viking may not handle such a code path as well as it could.

**Comment by rnorris** [comment]
I think I have understood the root cause now - which is present in the latest code.

Attached is a preview .patch file (although I had to rename the suffix to .md - in order to fix attaching said file - even though .patch files are meant to be accepted) - that should fix the segfault issue (and also the spurious glib/gtk warnings) that may of interest:

[WIP-Github-202-Increased-resilience-on-exit.md](https://github.com/viking-gps/viking/files/13326582/WIP-Github-202-Increased-resilience-on-exit.md)

Note this does not apply cleanly to the v1.10 - only to the latest HEAD.

I may get around to explaining this fix later... and of course mainlining properly in due course after more testing.

**Comment by rnorris** [comment]
So what was going wrong was:

Map downloading occurs in separate threads. When these threads finish, they may update some global data structures (the  list of requests [so we don't create multiple exact same requests] and the image map tile cache as result of these requests) - depending on the return status by libcurl that is performing the work on our behalf.

However when the program finishes these global structures get cleaned-up/de-allocated, and the download threads are terminated - if the thread is actually performing a download via libcurl, this termination results in an internal exit code - which was otherwise mishandled (i.e. not considered/known by Viking).

The previous patch fixes this mishandled status code and also includes protection against using the global data structures when they are not available.

**Comment by rnorris** [comment]
I don't think this is important enough to spend time creating an updated version 1.10 release.

It will simply be included when v1.11 is released.

**Comment by vinc17fr** [comment]
Thanks. I could port commit 21ecac202824cc91ddd7ce06ec01653e4cb025eb to viking 1.10 for Debian 12: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1055624#17

I've done several tests, and everything is now fine on my machine.
